crates/sui-core/src/safe_client.rs

```diff
@@ -581,4 +582,56 @@ where
         ));
         Ok(Box::pin(stream))
     }
+
+    fn verify_epoch(
+        &self,
+        requested_epoch_id: Option<EpochId>,
+        response: &EpochResponse,
+    ) -> SuiResult {
+        if let Some(epoch) = &response.epoch_info {
+            fp_ensure!(
+                requested_epoch_id.is_none() || requested_epoch_id == Some(epoch.epoch()),
+                SuiError::InvalidEpochResponse("Responded epoch number mismatch".to_string())
+            );
+        }
+        match (requested_epoch_id, &response.epoch_info) {
+            (None, None) => Err(SuiError::InvalidEpochResponse(
+                "Latest epoch must not be None".to_string(),
+            )),
+            (Some(epoch_id), None) => {
+                fp_ensure!(
+                    epoch_id != 0,
+                    SuiError::InvalidEpochResponse("Genesis epoch must be available".to_string())
+                );
+                Ok(())
+            }
+            (_, Some(AuthenticatedEpoch::Genesis(_))) => {
+                // TODO: Verify the epoch data using genesis committee
+                Ok(())
+            }
+            (_, Some(AuthenticatedEpoch::Signed(_))) => {
+                // TODO: Verify the epoch data using previous committee
+                Ok(())
+            }
+            (_, Some(AuthenticatedEpoch::Certified(_))) => {
+                // TODO: Verify the epoch data using previous committee
+                Ok(())
+            }
+        }
+    }
+
+    pub async fn handle_epoch(&self, request: EpochRequest) -> Result<EpochResponse, SuiError> {
+        let epoch_id = request.epoch_id;
+        let authority = self.address;
+        let response = self.authority_client.handle_epoch(request).await?;
+        self.verify_epoch(epoch_id, &response)
+            .map_err(|err| SuiError::ByzantineAuthoritySuspicion {
+                authority,
+                reason: err.to_string(),
+            })
+            .tap_err(|err| {
+                self.report_client_error(err);
+            })?;
```

Comment:
Generally this is not how logging should be handled, the macro invocation should be inlined here in order to properly report line numbers, etc.

Reply:
Good point

Reply:
Filed https://github.com/MystenLabs/sui/issues/3740